### ISSUE TITLE
Arguments of async `addFilter`

### ISSUE THREAD

**Issue by ArmorDarks** [mozilla/nunjucks#820]
It seems that async `addFilter`, on contrary to sync `addFilter`, can't handle more than one argument.

Consider this:

``` js
env.addFilter('lookup', function(name, callback) {
    db.getItem(name, callback);
}, true);
```

If we will try to pass more arguments, it won't work:

``` js
env.addFilter('lookup', function(name, type, preserve, props, callback) {
    db.getItem(name, callback);
}, true);
```

It will throw an error that `callback` isn't a function.

I suppose that happens because Nunjucks always providing as callback _second_ argument, while I think expected behavior is to always provide as a callback _last_ argument.

However, maybe it would be wiser to attach callback to `this` instead. In such case it would be enough to provide `true` as third parameter of `addFilter` and it will attach `this.resolve()` and `this.reject()` methods.

Another way is to follow Grunt's way: http://gruntjs.com/api/inside-tasks#this.async

It's quite cute, because you don't need to provide nor additional callback, nor `true` as third argument of `addFilter`. It's enough just to trigger `this.async()` to throw everything into async mode.

**Comment by maximus8891** [comment]
> It will throw an error that callback isn't a function.

I could not reproduce it. @ArmorDarks Can you show your template where you used it?
I want to see how many arguments you passed to the filter.

**Comment by ArmorDarks** [comment]
@maximus8891 Unfortunately I do not use async version of `addFilter` in none of my projects, so I can't give a better example than the one that already I wrote. Maybe it was fixed in 3.0.0?

Rolling back in history, this, I guess, is the branch at which I've used async addFilter last time: https://github.com/LotusTM/Kotsu/blob/4ec1dff99e718b0e14b4850c49e73f206ec51ae8/modules/nunjucks-extensions.coffee#L200

But it didn't use more than two arguments due to described in issue restriction, so it won't serve as a good test case anyway.

**Comment by maximus8891** [comment]
@ArmorDarks When I used an async filter as follows:
`{{ variable | asyncFilter(argument) }}`

I needed three arguments:
```javascript
env.addFilter('asyncFilter', function(variable, argument, callback) {
    db.getItem(variable, argument, callback);
}, true);
```

It used version 3.0.0

**Comment by ArmorDarks** [comment]
Does it work if you make it like so?

``` jinja2
env.addFilter('asyncFilter', function(variable, argument, options, callback) {
    db.getItem(variable, argument, callback);
}, true);
```

**Comment by maximus8891** [comment]
@ArmorDarks In this case `callback` would be undefined as `options` will be called as the callback.

**Comment by ArmorDarks** [comment]
@maximus8891 that's exactly what about this issue :)

**Comment by maximus8891** [comment]
@ArmorDarks I see. The only case I can think of when it is an issue is when you have optional arguments for a filter. For instance, when you want to make `{{ variable | asyncFilter(argument) }}` and `{{ variable | asyncFilter(argument, options) }}` valid. Remembering to provide `options` at all times solves it.

Looking at the code, the easiest way to fix it would be to change the order in which arguments are supplied in a compiled template:

```javascript
compileFilterAsync: function(node, frame) {
        var name = node.name;
        this.assertType(name, nodes.Symbol);

        var symbol = node.symbol.value;
        frame.set(symbol, symbol);

        this.emit('env.getFilter("' + name.value + '").call(context, ');
        this.emitLine(this.makeCallback(symbol) + ', ');        
        this._compileAggregate(node.args, frame);

        this.addScopeLevel();
    },
```

The filters would have to be defined like this:
```javascript
env.addFilter('asyncFilter', function(callback, variable, argument, options) {
    db.getItem(variable, argument, callback);
}, true);
```

**Comment by ArmorDarks** [comment]
@maximus8891 There are more elegant ways to do same thing. For example, to extract last argument as callback from `arguments` object which has every function by default.

Or, even better, to handle async with Grunt-like approach, which I've described in original issue.

**Comment by maximus8891** [comment]
The issue is not how to extract the callback function, but to know how many arguments the filter function accepts. Otherwise you can handle it in custom filters.
```js
env.addFilter('asyncFilter', function(/*variable, argument, options, callback*/) {
    const args = [].slice.call(arguments);
    const cb = args.pop();
    db.getItem.call(null, args[0], args[1], args[2], cb);
}, true);
```

**Comment by ArmorDarks** [comment]
When I was talking about more elegant ways, I meant how Nunjucks could injects callback function into async filters, not how would you get callback for your own filter.

So far I can't understand is it my description of issue so unclear, or it's just you getting whole thing completely wrong.

**Comment by RRMoelker** [comment]
I agree that Nunjucks should allow optional arguments without requiring a hack.

But for now, those using ES6, instead of `arguments`, try:

```
const myFilter = (...args) => {
  const cb = args.pop();
  const result = doStuff(args[0], args[1]); // etc
  cb(null, result);
}
```
